PerfTools/AllocMonitor/plugins/ModuleEventAllocMonitor.cc

```diff
@@ -0,0 +1,445 @@
+      return true;
+    }
+    return false;
+  }
+}  // namespace
+
+class ModuleEventAllocMonitor {
+  unsigned int moduleIndex(unsigned int mod_id) const {
+    auto index = mod_id;
+    if (not moduleIDs_.empty()) {
+      auto it = std::lower_bound(moduleIDs_.begin(), moduleIDs_.end(), mod_id);
+      index = it - moduleIDs_.begin();
+    }
+    return index;
+  }
+
+public:
+  ModuleEventAllocMonitor(edm::ParameterSet const& iPS, edm::ActivityRegistry& iAR)
+      : moduleNames_(iPS.getUntrackedParameter<std::vector<std::string>>("moduleNames")),
+        nEventsToSkip_(iPS.getUntrackedParameter<unsigned int>("nEventsToSkip")),
+        filter_(&moduleIDs_) {
+    (void)cms::perftools::AllocMonitorRegistry::instance().createAndRegisterMonitor<MonitorAdaptor>();
+
+    if (nEventsToSkip_ > 0) {
+      filter_.setGlobalKeep(false);
+    }
+    auto file = std::make_shared<edm::ThreadSafeOutputFileStream>(iPS.getUntrackedParameter<std::string>("fileName"));
+    {
+      std::stringstream s;
+      s << "#Format\n"
+           "# --------\n"
+           "# prefixes\n"
+           "# #: comment\n"
+           "# @: module info\n"
+           "# A: memory info for call to 'acquire'\n"
+           "# M: memory info for standard module method (i.e. produce, analyze or filter)\n"
+           "# D: memory reclaimed when Event products are being deleted at end of Event processing\n"
+           "# --------\n"
+           "# line formats\n"
+           "#@ <module label> <module type> <module ID>\n"
+           "#A <module ID> <stream #> <total temp memory (bytes)> <# temp allocations> <total unmatched deallocations "
+           "(bytes)> <# unmatched deallocations> <total unmatched allocations [this is copied to #M] (bytes)> <# "
+           "unmatched allocations [also copied]>\n"
+           "#M <module ID> <stream #> <total temp memory (bytes)> <# temp allocations> <total unmatched deallocations "
+           "(bytes)> <# unmatched deallocations> <total unmatched allocations (bytes)> <# unmatched allocations>\n"
+           "#D <module ID> <stream #> <total matched deallocations (bytes)> <# matched deallocations>\n";
+      file->write(s.str());
+    }
+    if (not moduleNames_.empty()) {
+      iAR.watchPreModuleConstruction([this, file](auto const& description) {
+        auto found = std::find(moduleNames_.begin(), moduleNames_.end(), description.moduleLabel());
+        if (found != moduleNames_.end()) {
+          moduleIDs_.push_back(description.id());
+          nModules_ = moduleIDs_.size();
+          std::sort(moduleIDs_.begin(), moduleIDs_.end());
+          std::stringstream s;
+          s << "@ " << description.moduleLabel() << " " << description.id() << "\n";
+          file->write(s.str());
+        }
+      });
+    }
+    if (nEventsToSkip_ > 0) {
+      iAR.watchPreSourceEvent([this](auto) {
+        ++nEventsStarted_;
+        if (nEventsStarted_ > nEventsToSkip_) {
+          filter_.setGlobalKeep(true);
```

Comment:
Does this mean that in a multi-stream job the data collection could start in the middle of the processing of an Event whose entry number is less than `nEventsToSkip_`?

Reply:
possibly

Reply:
Could this be documented somewhere (unless you plan to improve the behavior in this PR)?

Reply:
I added the caveat to the fillDescriptions comment.